src/bin/pg_autoctl/cli_create_drop_node.c

```diff
@@ -506,3 +559,130 @@ cli_drop_node(int argc, char **argv)
 
 	keeper_config_destroy(&config);
 }
+
+
+/*
+ * discover_nodename discovers a suitable --nodename default value in three
+ * steps:
+ *
+ * 1. First find the local LAN IP address by connecting a socket() to either an
+ *    internet service (8.8.8.8:53) or to the monitor's hostname and port, and
+ *    then inspecting which local address has been used.
+ *
+ * 2. Use the local IP address obtained in the first step and do a reverse DNS
+ *    lookup for it. The answer is our candidate default --nodename.
+ *
+ * 3. Do a DNS lookup for the candidate default --nodename. If we get back a IP
+ *    address that matches one of the local network interfaces, we keep the
+ *    candidate, the DNS lookup that Postgres does at connection time is
+ *    expected to then work.
+ *
+ * All this dansing around DNS lookups is necessary in order to mimic Postgres
+ * HBA matching of hostname rules against client IP addresses: the hostname in
+ * the HBA rule is resolved and compared to the client IP address. We want the
+ * --nodename we use to resolve to an IP address that exists on the local
+ * Postgres server.
+ *
+ * Worst case here is that we fail to discover a --nodename and then ask the
+ * user to provide one for us.
+ *
+ * monitorHostname and monitorPort are used to open a socket to that address,
+ * in order to find the right outbound interface. When creating a monitor node,
+ * of course, we don't have the monitorHostname yet: we are trying to discover
+ * it... in that case we use PG_AUTOCTL_DEFAULT_SERVICE_NAME and PORT, which
+ * are the Google DNS service: 8.8.8.8:53, expected to be reachable.
+ */
+static bool
+discover_nodename(char *nodename, int size,
+				  const char *monitorHostname, int monitorPort)
+{
+	/*
+	 * Try and find a default --nodename. The --nodename is mandatory, so
+	 * when not provided for by the user, then failure to discover a
+	 * suitable nodename is a fatal error.
+	 */
+	char ipAddr[BUFSIZE];
+	char localIpAddr[BUFSIZE];
+	char hostname[_POSIX_HOST_NAME_MAX];
+
+	/* fetch our local address among the network interfaces */
+	if (!fetchLocalIPAddress(ipAddr, BUFSIZE, monitorHostname, monitorPort))
+	{
+		log_fatal("Failed to find a local IP address, "
+				  "please provide --nodename.");
+		return false;
+	}
+
+	/* from there on we can take the ipAddr as the default --nodename */
+	strlcpy(nodename, ipAddr, size);
+	log_debug("check_or_discover_nodename: local ip %s", ipAddr);
```

Comment:
the first part of the logline seems to refer to a function name that is different from the one we are currently in. Is this a left over of a refactoring or by design?

Reply:
Left-over from refactoring. Good spot!